propolis/src/block/crucible.rs

```diff
@@ -0,0 +1,341 @@
+            sectors: 0,
+            read_only,
+            driver: Mutex::new(None),
+        };
+
+        // After up_main has returned successfully, wait for active negotiation
+        let uuid = guest.query_upstairs_uuid()?;
+
+        println!("Calling activate for {:?}", uuid);
+        guest.activate()?;
+
+        let mut active = false;
+        for _ in 0..10 {
+            if guest.query_is_active()? {
+                println!("{:?} is active", uuid);
+                active = true;
+                break;
+            }
+
+            std::thread::sleep(std::time::Duration::from_secs(2));
+        }
+        if !active {
+            return Err(crucible::CrucibleError::UpstairsInactive);
+        }
+
+        // After active negotiation, set sizes
+        be.block_size = guest.query_block_size()?;
+        be.total_size = guest.query_total_size()?;
+        be.sectors = be.total_size / be.block_size;
+
+        Ok(Arc::new(be))
+    }
+}
+
+impl block::Backend for CrucibleBackend {
+    fn info(&self) -> DeviceInfo {
+        DeviceInfo {
+            block_size: self.block_size as u32,
+            total_size: self.sectors as u64,
+            writable: !self.read_only,
+        }
+    }
+
+    fn attach(&self, dev: Arc<dyn block::Device>, disp: &Dispatcher) {
+        let mut driverg = self.driver.lock().unwrap();
+        assert!(driverg.is_none());
+
+        // spawn synchronous driver
+        let driver = SyncDriver::new(dev, self.guest.clone());
+        driver.spawn(NonZeroUsize::new(8).unwrap(), disp);
+        *driverg = Some(driver);
+    }
+}
+
+impl Entity for CrucibleBackend {}
+
+struct SyncDriver {
+    cv: Condvar,
+    queue: Mutex<VecDeque<block::Request>>,
+    idle_threads: Semaphore,
+    dev: Arc<dyn block::Device>,
+    guest: Arc<crucible::Guest>,
+    waiter: block::AsyncWaiter,
+}
+
+impl SyncDriver {
+    fn new(
+        dev: Arc<dyn block::Device>,
+        guest: Arc<crucible::Guest>,
+    ) -> Arc<Self> {
+        let waiter = block::AsyncWaiter::new(dev.as_ref());
+        Arc::new(Self {
+            cv: Condvar::new(),
+            queue: Mutex::new(VecDeque::new()),
+            idle_threads: Semaphore::new(0),
+            dev,
+            guest,
+            waiter,
+        })
+    }
+
+    fn blocking_loop(&self, sctx: &mut SyncCtx) {
+        loop {
+            if sctx.check_yield() {
+                break;
+            }
+
+            let mut guard = self.queue.lock().unwrap();
+            if let Some(req) = guard.pop_front() {
+                drop(guard);
+                let ctx = sctx.dispctx();
+                match process_request(self.guest.clone(), &req, &ctx) {
+                    Ok(_) => req.complete(block::Result::Success, &ctx),
+                    Err(_) => req.complete(block::Result::Failure, &ctx),
+                }
+            } else {
+                // wait until more requests are available
+                self.idle_threads.add_permits(1);
```

Comment:
As I found with some other code I was working on, this needs a guard, as a quiesce request can cause multiple trips over the wait condition, causing permits in excess of the thread count to accrue.

Reply:
I think it has to do with the wake function not resetting (or otherwise interacting) with the idle_threads semaphore, though I'm not sure what the right action to take is. I didn't have time to go further into this today.